### ISSUE TITLE
JavaScript file: Shadowing `require` still behaves like the global one

### ISSUE THREAD

**Issue by Derugon** [microsoft/TypeScript#61528]
### 🔎 Search Terms

require shadowing javascript import scope local

### 🕗 Version & Regression Information

- This changed between versions 4.0.5 and 4.1.2 (since `require` module resolution was introduced in 4.1.0-dev.20200818 with #39770).

### ⏯ Playground Link

https://www.typescriptlang.org/play/?esModuleInterop=false&jsx=0&module=0&filetype=js&q=422#code/GYVwdgxgLglg9mABAUwB4EMC2AHANsxACgEpEBvAKEWsQHoAqexAAW3QCctyBnKdmMAHMAvojBYC9WlRqhIsBInbIAjiBjLC4zMlJklyKCHZIAciEwAjZOy0TSwmdQgJeibnB3mrNxAF4DNQ1kQgByDy8JUOIAbgphIA

### 💻 Code

```js
function () {
    /** @param {string} name */
    function require(name) { return Number(name) }

    const someNumber = require('someName');
}
```

### 🙁 Actual behavior

1. `someNumber` is an import.
2. It raises a type import error:
```
Cannot find module 'someName' or its corresponding type declarations.
```

### 🙂 Expected behavior

1. `someNumber` has type `number`.
2. No error is raised.

### Additional information about the issue

This is only the case when using type inference with JavaScript files. With TypeScript it works as expected ([playground](https://www.typescriptlang.org/play/?esModuleInterop=false&jsx=0&module=0&ssl=4&ssc=2&pln=1&pc=1&filetype=ts&q=422#code/GYVwdgxgLglg9mABAUwB4EMC2AHANsxACgEpEBvAKEWsVElgUQCdkBHEGFwsLZALkQBnKExhgA5qTLNkUEEyQA5EJgBGyJt16kAvlRoQEwoXEzJlajYgC8M9p2SEA5INPneT4gG4KOoA)):
```ts
function () {
    function require(name: string) { return Number(name) }

    const someNumber = require('someName');
}
```
If the type of `someNumber` is explicitely written, it also works as expected ([playground](https://www.typescriptlang.org/play/?esModuleInterop=false&jsx=0&module=0&filetype=js&q=422#code/GYVwdgxgLglg9mABAUwB4EMC2AHANsxACgEpEBvAKEWsQHoAqexAAW3QCctyBnKdmMAHMAvojBYC9WlRqhIsBInbIAjiBjLC4zMlJklyKCHZIAciEwAjZOy0TSwmdQZNmUAJ7YCZMBevtRKSdECAReRG44HXMrG0QAXgM1DWRCAHJI6Ik04gBuCmEgA)):
```js
function () {
    /** @param {string} name */
    function require(name) { return Number(name) }

    /** @type {number} */
    const someNumber = require('someName');
}
```

**Comment by RyanCavanaugh** [comment]
There are too many other cases where `require` appears shadowed but actually isn't (e.g. AMD module bodies). For consistency, `require` is always inferred to be the module-importing function.
